File: src/handlers/lib/users/update/email.rs
use crate::{
    config::Config,
    crypto::{jwt::JWT, secure_token},
    diesel::Connection,
    handlers::Error,
    mailer::send_email,
    models::user::User,
};
use chrono::Utc;
use diesel::{
    pg::PgConnection,
    r2d2::{ConnectionManager, PooledConnection},
};
use log::error;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, GraphQLInputObject)]
pub struct UpdateForm {
    pub email: String,
    pub confirm: Option<bool>,
}

pub fn update_email(config: &Config, connection: &PooledConnection<ConnectionManager<PgConnection>>, token: &JWT, update_form: UpdateForm, id: String) -> Result<User, Error> {
    if !token.is_admin(connection) {
        return Err(Error::new(403, json!({"code": "only_admin_can_update"}), "Only Admin Can Update Users".to_string()));
    }

    let user = crate::models::user::get_by_id(id, connection);

    if user.is_err() {
        let err = user.err().unwrap();

        error!("{:?}", err);

        return Err(Error::from(err));
    }

    let mut user = user.unwrap();

    if user.id == token.sub {
        return Err(Error::new(422, json!({"code": "admin_cant_update_self"}), "Admin Can't Update Self".to_string()));
    }

    if config.auto_confirm || update_form.confirm.is_some() && update_form.confirm.unwrap() {
        user.new_email = user.email.clone(); // store the old email in new email in case we ever need to revert it

        user.email = Some(update_form.email);

        let user = user.save(connection);

        if user.is_err() {
            let err = user.err().unwrap();

            error!("{:?}", err);

            return Err(Error::from(err));
        }

        return Ok(user.unwrap());
    }

    let transaction = connection.transaction::<User, Error, _>(|| {
        user.new_email = Some(update_form.email.clone());

        user.email_change_token = Some(secure_token(100));

        user.email_change_token_sent_at = Some(Utc::now().naive_utc());

        let user = user.save(&connection);

        if user.is_err() {
            let err = user.err().unwrap();

            error!("{:?}", err);

            return Err(Error::from(err));
        }

        let user = user.unwrap();

        let template = &config.get_change_email_template();

        let to = &user.new_email.unwrap();

        let subject = &config.get_confirmation_email_subject();

        let data = json!({
            "change_email_token": user.email_change_token.clone().unwrap(),
            "email": user.email,
            "new_email": user.new_email,
            "site_url": config.site_url
        });

        let email = send_email(template, data, to, subject, config);

        if email.is_err() {
            let err = email.err().unwrap();

            error!("{:?}", err);

            return Err(Error::from(err));
        }

        return Ok(user);
    });

    if transaction.is_err() {
        return Err(transaction.err().unwrap());
    }

    return Ok(transaction.unwrap());
}

File: handlers/lib/users/update/email.go
package update

import (
	"github.com/sirupsen/logrus"
	"github.com/thanhpk/randstr"
	"github.com/zolamk/trust/config"
	"github.com/zolamk/trust/handlers"
	"github.com/zolamk/trust/jwt"
	"github.com/zolamk/trust/lib/email"
	"github.com/zolamk/trust/model"
	"gorm.io/gorm"
)

func UpdateEmail(db *gorm.DB, config *config.Config, token *jwt.JWT, id string, new_email string, confirm *bool) (*model.User, error) {

	is_admin, err := token.IsAdmin(db)

	if err != nil {
		logrus.Error(err)
		return nil, handlers.ErrInternal
	}

	if !is_admin {
		return nil, handlers.ErrAdminOnly
	}

	if token.Subject == id {
		return nil, handlers.ErrCantChangeOwnAccount
	}

	user := &model.User{}

	tx := db.First(user, "email = ?", new_email)

	if tx.Error != nil {

		if tx.Error != gorm.ErrRecordNotFound {
			logrus.Error(tx.Error)
			return nil, handlers.ErrInternal
		}

	} else {

		err := handlers.ErrEmailRegistered

		err.Extensions["email"] = user.Email

		err.Extensions["phone"] = user.Phone

		err.Extensions["id"] = user.ID

		return nil, err

	}

	if tx := db.Find(user, "id = ?", id); tx.Error != nil {
		if tx.Error == gorm.ErrRecordNotFound {
			return nil, handlers.ErrUserNotFound
		}
		logrus.Error(tx.Error)
		return nil, handlers.ErrInternal
	}

	if !config.EmailRule.MatchString(new_email) {
		return nil, handlers.ErrInvalidEmail
	}

	err = db.Transaction(func(tx *gorm.DB) error {

		user.NewEmail = &new_email

		if err := user.Save(db); err != nil {
			logrus.Error(err)
			return handlers.ErrInternal
		}

		if config.AutoConfirm || (confirm != nil && *confirm) {

			if err := user.ConfirmEmailChange(tx); err != nil {
				logrus.Error(err)
				return handlers.ErrInternal
			}

			return nil

		}

		token := randstr.String(100)

		user.EmailChangeToken = &token

		user.NewEmail = &new_email

		context := &map[string]string{
			"site_url":           config.SiteURL,
			"email_change_token": *user.EmailChangeToken,
			"new_email":          *user.NewEmail,
			"instance_url":       config.InstanceURL,
		}

		if err := email.SendEmail(config.ChangeTemplate, context, user.NewEmail, config); err != nil {
			logrus.Error(err)
			return handlers.ErrInternal
		}

		return nil

	})

	if err != nil {
		return nil, err
	}

	return user, nil

}
